File: scripts/agent-runtime/scenario-results.mjs
import {
	COMPARE_ARTIFACT_SCHEMA,
	SCENARIO_RESULTS_SCHEMA,
} from "./contract-versions.mjs";

const MODE_VALUES = new Set(["iterate", "held_out", "comparison", "full_gate"]);
const RESULT_STATUS_VALUES = new Set(["passed", "failed", "improved", "regressed", "unchanged", "noisy", "missing"]);
const COMPARE_VERDICT_VALUES = new Set(["improved", "regressed", "mixed", "unchanged", "inconclusive"]);
const COMPARE_DELTA_STATUS_VALUES = new Set(["improved", "regressed", "unchanged", "noisy"]);
const TELEMETRY_NUMERIC_FIELDS = ["prompt_tokens", "completion_tokens", "total_tokens", "cost_usd"];

function normalizeNonEmptyString(value, field) {
	if (typeof value !== "string" || !value.trim()) {
		throw new Error(`${field} must be a non-empty string`);
	}
	return value;
}

function normalizeOptionalString(value, field) {
	if (value === undefined || value === null) {
		return undefined;
	}
	return normalizeNonEmptyString(value, field);
}

function normalizeIsoTimestamp(value, field) {
	if (value === undefined || value === null) {
		return undefined;
	}
	if (typeof value !== "string" || !Number.isFinite(Date.parse(value))) {
		throw new Error(`${field} must be a valid ISO timestamp`);
	}
	return value;
}

function normalizeNonNegativeNumber(value, field) {
	if (value === undefined || value === null) {
		return undefined;
	}
	if (typeof value !== "number" || !Number.isFinite(value) || value < 0) {
		throw new Error(`${field} must be a non-negative number`);
	}
	return value;
}

function normalizeResultStatus(value, field) {
	if (value === undefined || value === null) {
		return undefined;
	}
	const status = normalizeNonEmptyString(value, field);
	if (!RESULT_STATUS_VALUES.has(status)) {
		throw new Error(`${field} must be one of ${[...RESULT_STATUS_VALUES].join(", ")}`);
	}
	return status;
}

function normalizeMode(value, field) {
	if (value === undefined || value === null) {
		return undefined;
	}
	const mode = normalizeNonEmptyString(value, field);
	if (!MODE_VALUES.has(mode)) {
		throw new Error(`${field} must be one of ${[...MODE_VALUES].join(", ")}`);
	}
	return mode;
}

function normalizeTelemetry(value, field) {
	if (value === undefined || value === null) {
		return undefined;
	}
	if (!value || typeof value !== "object" || Array.isArray(value)) {
		throw new Error(`${field} must be an object`);
	}
	const telemetry = {};
	for (const key of ["provider", "model"]) {
		const normalized = normalizeOptionalString(value[key], `${field}.${key}`);
		if (normalized !== undefined) {
			telemetry[key] = normalized;
		}
	}
	for (const key of TELEMETRY_NUMERIC_FIELDS) {
		const normalized = normalizeNonNegativeNumber(value[key], `${field}.${key}`);
		if (normalized !== undefined) {
			telemetry[key] = normalized;
		}
	}
	return Object.keys(telemetry).length > 0 ? telemetry : undefined;
}

function assertArray(value, field) {
	if (!Array.isArray(value)) {
		throw new Error(`${field} must be an array`);
	}
	return value;
}

function normalizeScenarioBucketEntry(value, field) {
	if (typeof value === "string") {
		return normalizeNonEmptyString(value, field);
	}
	if (!value || typeof value !== "object" || Array.isArray(value)) {
		throw new Error(`${field} must be a string or object`);
	}
	if (value.scenarioId !== undefined) {
		normalizeNonEmptyString(value.scenarioId, `${field}.scenarioId`);
	}
	if (value.metric !== undefined) {
		normalizeNonEmptyString(value.metric, `${field}.metric`);
	}
	if (value.reason !== undefined) {
		normalizeNonEmptyString(value.reason, `${field}.reason`);
	}
	if (value.summary !== undefined) {
		normalizeNonEmptyString(value.summary, `${field}.summary`);
	}
	return value;
}

function normalizeScenarioBucket(value, field) {
	if (value === undefined || value === null) {
		return [];
	}
	return assertArray(value, field).map((entry, index) =>
		normalizeScenarioBucketEntry(entry, `${field}[${index}]`),
	);
}

function normalizeCompareDelta(value, index, field) {
	if (!value || typeof value !== "object" || Array.isArray(value)) {
		throw new Error(`${field}[${index}] must be an object`);
	}
	const status = normalizeNonEmptyString(value.status, `${field}[${index}].status`);
	if (!COMPARE_DELTA_STATUS_VALUES.has(status)) {
		throw new Error(`${field}[${index}].status must be one of ${[...COMPARE_DELTA_STATUS_VALUES].join(", ")}`);
	}
	const normalized = {
		key: normalizeNonEmptyString(value.key, `${field}[${index}].key`),
		status,
		summary: normalizeNonEmptyString(value.summary, `${field}[${index}].summary`),
	};
	for (const optionalField of ["scenarioId", "metric", "artifactPath"]) {
		const optionalValue = normalizeOptionalString(
			value[optionalField],
			`${field}[${index}].${optionalField}`,
		);
		if (optionalValue !== undefined) {
			normalized[optionalField] = optionalValue;
		}
	}
	for (const numericField of ["baselineValue", "candidateValue"]) {
		const numericValue = normalizeNonNegativeNumber(
			value[numericField],
			`${field}[${index}].${numericField}`,
		);
		if (numericValue !== undefined) {
			normalized[numericField] = numericValue;
		}
	}
	return normalized;
}

function normalizeCompareVerdict(value, field) {
	const verdict = normalizeOptionalString(value, field);
	if (verdict !== undefined && !COMPARE_VERDICT_VALUES.has(verdict)) {
		throw new Error(`${field} must be one of ${[...COMPARE_VERDICT_VALUES].join(", ")}`);
	}
	return verdict;
}

function addCompareBuckets(normalized, value, field) {
	for (const bucketField of ["improved", "regressed", "unchanged", "noisy"]) {
		const bucket = normalizeScenarioBucket(value[bucketField], `${field}.${bucketField}`);
		if (bucket.length > 0) {
			normalized[bucketField] = bucket;
		}
	}
}

function addCompareDeltas(normalized, value, field) {
	const deltas = value.deltas === undefined ? [] : assertArray(value.deltas, `${field}.deltas`);
	if (deltas.length > 0) {
		normalized.deltas = deltas.map((entry, index) => normalizeCompareDelta(entry, index, `${field}.deltas`));
	}
}

function addCompareArtifactPaths(normalized, value, field) {
	const artifactPaths = value.artifactPaths === undefined ? [] : assertArray(value.artifactPaths, `${field}.artifactPaths`);
	if (artifactPaths.length > 0) {
		normalized.artifactPaths = artifactPaths.map((entry, index) =>
			normalizeNonEmptyString(entry, `${field}.artifactPaths[${index}]`),
		);
	}
}

export function normalizeCompareArtifact(value, field = "compareArtifact") {
	if (value === undefined || value === null) {
		return undefined;
	}
	if (!value || typeof value !== "object" || Array.isArray(value)) {
		throw new Error(`${field} must be an object`);
	}
	if (value.schemaVersion !== COMPARE_ARTIFACT_SCHEMA) {
		throw new Error(`${field}.schemaVersion must be ${COMPARE_ARTIFACT_SCHEMA}`);
	}
	const verdict = normalizeCompareVerdict(value.verdict, `${field}.verdict`);
	const normalized = {
		schemaVersion: COMPARE_ARTIFACT_SCHEMA,
		summary: normalizeNonEmptyString(value.summary, `${field}.summary`),
	};
	const generatedAt = normalizeIsoTimestamp(value.generatedAt, `${field}.generatedAt`);
	if (generatedAt !== undefined) {
		normalized.generatedAt = generatedAt;
	}
	if (verdict !== undefined) {
		normalized.verdict = verdict;
	}
	addCompareBuckets(normalized, value, field);
	addCompareDeltas(normalized, value, field);
	addCompareArtifactPaths(normalized, value, field);
	return normalized;
}

function addScenarioResultMetrics(normalized, result, label) {
	const overallScore = normalizeNonNegativeNumber(result.overallScore, `${label}.overallScore`);
	if (overallScore !== undefined) {
		normalized.overallScore = overallScore;
	}
	const passRate = normalizeNonNegativeNumber(result.passRate, `${label}.passRate`);
	if (passRate !== undefined) {
		if (passRate > 1) {
			throw new Error(`${label}.passRate must be <= 1`);
		}
		normalized.passRate = passRate;
	}
}

function addScenarioResultTimes(normalized, result, label) {
	for (const field of ["timestamp", "startedAt", "completedAt"]) {
		const normalizedValue = normalizeIsoTimestamp(result[field], `${label}.${field}`);
		if (normalizedValue !== undefined) {
			normalized[field] = normalizedValue;
		}
	}
	const durationMs = normalizeNonNegativeNumber(result.durationMs, `${label}.durationMs`);
	if (durationMs !== undefined) {
		normalized.durationMs = durationMs;
	}
}

export function normalizeScenarioResult(result, label = "result") {
	if (!result || typeof result !== "object" || Array.isArray(result)) {
		throw new Error(`${label} must be an object`);
	}
	const normalized = {
		scenarioId: normalizeNonEmptyString(result.scenarioId, `${label}.scenarioId`),
	};
	const status = normalizeResultStatus(result.status, `${label}.status`);
	if (status !== undefined) {
		normalized.status = status;
	}
	addScenarioResultMetrics(normalized, result, label);
	addScenarioResultTimes(normalized, result, label);
	const telemetry = normalizeTelemetry(result.telemetry, `${label}.telemetry`);
	if (telemetry !== undefined) {
		normalized.telemetry = telemetry;
	}
	return normalized;
}

export function normalizeScenarioResultsPacket(value, field = "scenarioResults") {
	if (!value || typeof value !== "object" || Array.isArray(value)) {
		throw new Error(`${field} must be an object`);
	}
	if (value.schemaVersion !== SCENARIO_RESULTS_SCHEMA) {
		throw new Error(`${field}.schemaVersion must be ${SCENARIO_RESULTS_SCHEMA}`);
	}
	const normalized = {
		schemaVersion: SCENARIO_RESULTS_SCHEMA,
		results: assertArray(value.results, `${field}.results`).map((entry, index) =>
			normalizeScenarioResult(entry, `${field}.results[${index}]`),
		),
	};
	const source = normalizeOptionalString(value.source, `${field}.source`);
	if (source !== undefined) {
		normalized.source = source;
	}
	const generatedAt = normalizeIsoTimestamp(value.generatedAt, `${field}.generatedAt`);
	if (generatedAt !== undefined) {
		normalized.generatedAt = generatedAt;
	}
	const mode = normalizeMode(value.mode, `${field}.mode`);
	if (mode !== undefined) {
		normalized.mode = mode;
	}
	const compareArtifact = normalizeCompareArtifact(value.compareArtifact, `${field}.compareArtifact`);
	if (compareArtifact !== undefined) {
		normalized.compareArtifact = compareArtifact;
	}
	return normalized;
}

File: internal/runtime/scenario_results.go
package runtime

import (
	"fmt"
	"sort"
	"strings"

	"github.com/corca-ai/cautilus/internal/contracts"
)

var (
	modeValues         = map[string]struct{}{"iterate": {}, "held_out": {}, "comparison": {}, "full_gate": {}}
	resultStatusValues = map[string]struct{}{
		"passed": {}, "failed": {}, "improved": {}, "regressed": {}, "unchanged": {}, "noisy": {}, "missing": {},
	}
	compareVerdictValues     = map[string]struct{}{"improved": {}, "regressed": {}, "mixed": {}, "unchanged": {}, "inconclusive": {}}
	compareDeltaStatusValues = map[string]struct{}{"improved": {}, "regressed": {}, "unchanged": {}, "noisy": {}}
	telemetryNumericFields   = []string{"prompt_tokens", "completion_tokens", "total_tokens", "cost_usd"}
)

func normalizeCompareArtifact(value any, field string) (map[string]any, error) {
	if value == nil {
		return nil, nil
	}
	record, ok := value.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("%s must be an object", field)
	}
	if record["schemaVersion"] != contracts.CompareArtifactSchema {
		return nil, fmt.Errorf("%s.schemaVersion must be %s", field, contracts.CompareArtifactSchema)
	}
	summary, err := normalizeNonEmptyString(record["summary"], field+".summary")
	if err != nil {
		return nil, err
	}
	normalized := map[string]any{
		"schemaVersion": contracts.CompareArtifactSchema,
		"summary":       summary,
	}
	if generatedAt, err := normalizeISOTimestamp(record["generatedAt"], field+".generatedAt"); err != nil {
		return nil, err
	} else if generatedAt != nil {
		normalized["generatedAt"] = *generatedAt
	}
	if verdict, err := normalizeOptionalString(record["verdict"], field+".verdict"); err != nil {
		return nil, err
	} else if verdict != nil {
		if _, ok := compareVerdictValues[*verdict]; !ok {
			return nil, fmt.Errorf("%s.verdict must be one of %s", field, strings.Join(sortedKeys(compareVerdictValues), ", "))
		}
		normalized["verdict"] = *verdict
	}
	for _, bucketField := range []string{"improved", "regressed", "unchanged", "noisy"} {
		bucket, err := normalizeScenarioBucket(record[bucketField], field+"."+bucketField)
		if err != nil {
			return nil, err
		}
		if len(bucket) > 0 {
			normalized[bucketField] = bucket
		}
	}
	if rawDeltas, err := assertArray(record["deltas"], field+".deltas"); err != nil {
		return nil, err
	} else if len(rawDeltas) > 0 {
		deltas := make([]any, 0, len(rawDeltas))
		for index, entry := range rawDeltas {
			delta, err := normalizeCompareDelta(entry, index, field+".deltas")
			if err != nil {
				return nil, err
			}
			deltas = append(deltas, delta)
		}
		normalized["deltas"] = deltas
	}
	if rawReasons, err := assertArray(record["reasons"], field+".reasons"); err != nil {
		return nil, err
	} else if len(rawReasons) > 0 {
		reasons, err := stringSliceFromAny(rawReasons, field+".reasons")
		if err != nil {
			return nil, err
		}
		normalized["reasons"] = reasons
	}
	if rawPaths, err := assertArray(record["artifactPaths"], field+".artifactPaths"); err != nil {
		return nil, err
	} else if len(rawPaths) > 0 {
		paths, err := stringSliceFromAny(rawPaths, field+".artifactPaths")
		if err != nil {
			return nil, err
		}
		normalized["artifactPaths"] = paths
	}
	return normalized, nil
}

func normalizeCompareDelta(value any, index int, field string) (map[string]any, error) {
	record, ok := value.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("%s[%d] must be an object", field, index)
	}
	status, err := normalizeNonEmptyString(record["status"], fmt.Sprintf("%s[%d].status", field, index))
	if err != nil {
		return nil, err
	}
	if _, ok := compareDeltaStatusValues[status]; !ok {
		return nil, fmt.Errorf("%s[%d].status must be one of %s", field, index, strings.Join(sortedKeys(compareDeltaStatusValues), ", "))
	}
	key, err := normalizeNonEmptyString(record["key"], fmt.Sprintf("%s[%d].key", field, index))
	if err != nil {
		return nil, err
	}
	summary, err := normalizeNonEmptyString(record["summary"], fmt.Sprintf("%s[%d].summary", field, index))
	if err != nil {
		return nil, err
	}
	normalized := map[string]any{
		"key":     key,
		"status":  status,
		"summary": summary,
	}
	for _, optionalField := range []string{"scenarioId", "metric", "artifactPath"} {
		value, err := normalizeOptionalString(record[optionalField], fmt.Sprintf("%s[%d].%s", field, index, optionalField))
		if err != nil {
			return nil, err
		}
		if value != nil {
			normalized[optionalField] = *value
		}
	}
	for _, numericField := range []string{"baselineValue", "candidateValue"} {
		value, err := normalizeNonNegativeNumber(record[numericField], fmt.Sprintf("%s[%d].%s", field, index, numericField))
		if err != nil {
			return nil, err
		}
		if value != nil {
			normalized[numericField] = *value
		}
	}
	return normalized, nil
}

func normalizeScenarioBucket(value any, field string) ([]any, error) {
	if value == nil {
		return []any{}, nil
	}
	items, err := assertArray(value, field)
	if err != nil {
		return nil, err
	}
	result := make([]any, 0, len(items))
	for index, entry := range items {
		if text, ok := entry.(string); ok {
			normalized, err := normalizeNonEmptyString(text, fmt.Sprintf("%s[%d]", field, index))
			if err != nil {
				return nil, err
			}
			result = append(result, normalized)
			continue
		}
		record, ok := entry.(map[string]any)
		if !ok {
			return nil, fmt.Errorf("%s[%d] must be a string or object", field, index)
		}
		if scenarioID, ok := record["scenarioId"]; ok && scenarioID != nil {
			if _, err := normalizeNonEmptyString(scenarioID, fmt.Sprintf("%s[%d].scenarioId", field, index)); err != nil {
				return nil, err
			}
		}
		if metric, ok := record["metric"]; ok && metric != nil {
			if _, err := normalizeNonEmptyString(metric, fmt.Sprintf("%s[%d].metric", field, index)); err != nil {
				return nil, err
			}
		}
		if reason, ok := record["reason"]; ok && reason != nil {
			if _, err := normalizeNonEmptyString(reason, fmt.Sprintf("%s[%d].reason", field, index)); err != nil {
				return nil, err
			}
		}
		if summary, ok := record["summary"]; ok && summary != nil {
			if _, err := normalizeNonEmptyString(summary, fmt.Sprintf("%s[%d].summary", field, index)); err != nil {
				return nil, err
			}
		}
		result = append(result, record)
	}
	return result, nil
}

func normalizeScenarioTelemetry(value any, field string) (map[string]any, error) {
	if value == nil {
		return nil, nil
	}
	record, ok := value.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("%s must be an object", field)
	}
	telemetry := map[string]any{}
	for _, key := range []string{"runtime", "provider", "model", "resolved_model", "model_revision", "session_mode", "cost_truth", "pricing_source", "pricing_version", "source"} {
		value, err := normalizeOptionalString(record[key], field+"."+key)
		if err != nil {
			return nil, err
		}
		if value != nil {
			telemetry[key] = *value
		}
	}
	fingerprint, err := normalizeRuntimeFingerprint(record["runtimeFingerprint"], record, field+".runtimeFingerprint")
	if err != nil {
		return nil, err
	}
	if len(fingerprint) > 0 {
		telemetry["runtimeFingerprint"] = fingerprint
	}
	for _, key := range telemetryNumericFields {
		value, err := normalizeNonNegativeNumber(record[key], field+"."+key)
		if err != nil {
			return nil, err
		}
		if value != nil {
			telemetry[key] = *value
		}
	}
	if len(telemetry) == 0 {
		return nil, nil
	}
	return telemetry, nil
}

func normalizeScenarioResult(result any, field string) (map[string]any, error) {
	record, ok := result.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("%s must be an object", field)
	}
	scenarioID, err := normalizeNonEmptyString(record["scenarioId"], field+".scenarioId")
	if err != nil {
		return nil, err
	}
	normalized := map[string]any{
		"scenarioId": scenarioID,
	}
	if status, err := normalizeOptionalString(record["status"], field+".status"); err != nil {
		return nil, err
	} else if status != nil {
		if _, ok := resultStatusValues[*status]; !ok {
			return nil, fmt.Errorf("%s.status must be one of %s", field, strings.Join(sortedKeys(resultStatusValues), ", "))
		}
		normalized["status"] = *status
	}
	if overallScore, err := normalizeNonNegativeNumber(record["overallScore"], field+".overallScore"); err != nil {
		return nil, err
	} else if overallScore != nil {
		normalized["overallScore"] = *overallScore
	}
	if passRate, err := normalizeNonNegativeNumber(record["passRate"], field+".passRate"); err != nil {
		return nil, err
	} else if passRate != nil {
		if *passRate > 1 {
			return nil, fmt.Errorf("%s.passRate must be <= 1", field)
		}
		normalized["passRate"] = *passRate
	}
	for _, timeField := range []string{"timestamp", "startedAt", "completedAt"} {
		value, err := normalizeISOTimestamp(record[timeField], field+"."+timeField)
		if err != nil {
			return nil, err
		}
		if value != nil {
			normalized[timeField] = *value
		}
	}
	if durationMs, err := normalizeNonNegativeNumber(record["durationMs"], field+".durationMs"); err != nil {
		return nil, err
	} else if durationMs != nil {
		normalized["durationMs"] = *durationMs
	}
	if telemetry, err := normalizeScenarioTelemetry(record["telemetry"], field+".telemetry"); err != nil {
		return nil, err
	} else if telemetry != nil {
		normalized["telemetry"] = telemetry
	}
	return normalized, nil
}

func NormalizeScenarioResultsPacket(value any, field string) (map[string]any, error) {
	record, ok := value.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("%s must be an object", field)
	}
	if record["schemaVersion"] != contracts.ScenarioResultsSchema {
		return nil, fmt.Errorf("%s.schemaVersion must be %s", field, contracts.ScenarioResultsSchema)
	}
	rawResults, err := assertArray(record["results"], field+".results")
	if err != nil {
		return nil, err
	}
	results := make([]any, 0, len(rawResults))
	for index, entry := range rawResults {
		normalized, err := normalizeScenarioResult(entry, fmt.Sprintf("%s.results[%d]", field, index))
		if err != nil {
			return nil, err
		}
		results = append(results, normalized)
	}
	normalized := map[string]any{
		"schemaVersion": contracts.ScenarioResultsSchema,
		"results":       results,
	}
	if source, err := normalizeOptionalString(record["source"], field+".source"); err != nil {
		return nil, err
	} else if source != nil {
		normalized["source"] = *source
	}
	if generatedAt, err := normalizeISOTimestamp(record["generatedAt"], field+".generatedAt"); err != nil {
		return nil, err
	} else if generatedAt != nil {
		normalized["generatedAt"] = *generatedAt
	}
	if mode, err := normalizeOptionalString(record["mode"], field+".mode"); err != nil {
		return nil, err
	} else if mode != nil {
		if _, ok := modeValues[*mode]; !ok {
			return nil, fmt.Errorf("%s.mode must be one of %s", field, strings.Join(sortedKeys(modeValues), ", "))
		}
		normalized["mode"] = *mode
	}
	if compareArtifact, err := normalizeCompareArtifact(record["compareArtifact"], field+".compareArtifact"); err != nil {
		return nil, err
	} else if compareArtifact != nil {
		normalized["compareArtifact"] = compareArtifact
	}
	return normalized, nil
}

func sortedKeys(values map[string]struct{}) []string {
	keys := make([]string, 0, len(values))
	for key := range values {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	return keys
}
